### PR TITLE
🌱 types_apiexport: make all and resourceSelector/namespace optional

### PR DESCRIPTION
**Pull request by s-urbaniak**
## Summary
This sets the fields `all` and `resourceSelector/namespace` optional.

## Related issue(s)

Fixes https://github.com/kcp-dev/kcp/pull/2222#discussion_r1008530913
Fixes https://github.com/kcp-dev/kcp/pull/2222#discussion_r1008532082

### CODE REVIEW

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 1]
```diff
@@ -193,9 +193,9 @@ type PermissionClaim struct {
 
 	// all claims all resources for the given group/resource.
 	// This is mutually exclusive with resourceSelector.
-	// +required
+	// +optional
 	// +kubebuilder:default=false
-	All bool `json:"all"`
+	All bool `json:"all,omitempty"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:198]: was originally expecting this as part fo the ResourceSelector.

Related question: what does an empty, but set `resourceSelector` mean?

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: see my question here: https://github.com/kcp-dev/kcp/pull/2222#discussion_r1009281781

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: If we remove the required tag, this leaves space for new field but does not check invariants. The only thing i can think of is adding a CEL validation here?

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: added CEL validation, does this make things look better?

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 2]
```diff
@@ -210,6 +210,8 @@ type PermissionClaim struct {
 	IdentityHash string `json:"identityHash,omitempty"`
 }
 
+// TODO(sur): change the expression to "has(self.name) != has(self.__namespace__)" once https://github.com/google/cel-go/pull/571 is availabe in kcp/kubernetes (fixed in k8s 1.25).
+// +kubebuilder:validation:XValidation:rule="has(self.name) && !has(self.__namespace__) || !has(self.name) && has(self.__namespace__)",message="either \"name\" or \"namespace\" must be set"
```
Comment (s-urbaniak) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:214]: note to reviewers: currently there is an unfortunate bug in 1.24's CEL.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: > was originally expecting this as part fo the ResourceSelector.

clarified out of band: this is fine, as resource selector is a slice.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go]: resolved, we don't need xor here

**Comment by s-urbaniak** [comment]
found a quirk in API semantics, holding

**Comment by nrb** [comment]
@s-urbaniak Ideally we'd make `All` and `ResourceSelector` mutually exclusive and one of them required, but Go type semantics make this a little frustrating.

I like keeping `All` explicit, because it's giving wide-ranging access. Accidentally omitting a resource selector could mean exposure of all `Secrets`, as an example. While that's not necessarily different from the current state of affairs in Kubernetes, I still think it's not a state we want to be in.

Could we use a bool pointer for `All` and validate that either `All` or `ResourceSelector` are non-nil (but not both)?

**Comment by s-urbaniak** [comment]
unsetting WIP as the state of this PR indeed reflects what we discussed :-)

File: config/crds/apis.kcp.dev_apibindings.yaml [hunk 3]
```diff
@@ -92,17 +92,18 @@ spec:
                               metadata.namespace field. If "name" is unset, all objects
                               from the namespace are being claimed.
                             type: string
-                        required:
-                        - namespace
                         type: object
+                        x-kubernetes-validations:
+                        - message: either "namespace" or "name" and "namespace" must
+                            be set
+                          rule: has(self.__namespace__) || has(self.name) && has(self.__namespace__)
```
Comment (sttts) [inline config/crds/apis.kcp.dev_apibindings.yaml:99]: isn't it the same as "namespace must be set" ? `A or (A and B)` is the same as `A`.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 1 above)
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: is there a check that either `All` is true or `len(selectors)>0`?

If I am not missing anything, an empty resource selector and `All==false` is valid now.

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: The rule `self.all != has(self.resourceSelector)` covers that, unit test: https://github.com/kcp-dev/kcp/blob/831cc1a5395cde8de761c8c0745977fef6f7adbf/pkg/apis/apis/v1alpha1/types_apiexport_test.go#L60-L68

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: fixed

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 4]
```diff
@@ -210,9 +210,11 @@ type PermissionClaim struct {
 	IdentityHash string `json:"identityHash,omitempty"`
 }
 
+// +kubebuilder:validation:XValidation:rule="has(self.__namespace__) || has(self.name)",message="at least one field must be set"
```
Comment (s-urbaniak) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:213]: @sttts `self != {}` won't work, I get CEl compilation errors.

File: config/crds/apis.kcp.dev_apibindings.yaml (hunk 3 above)
Comment (s-urbaniak) [reply config/crds/apis.kcp.dev_apibindings.yaml]: fixed, using a different expression to allow both `name` and `namespace`

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 5]
```diff
@@ -210,20 +209,28 @@ type PermissionClaim struct {
...
 type ResourceSelector struct {
 	// name of an object within a claimed group/resource.
 	// It matches the metadata.name field of the underlying object.
+	// If namespace is unset, all objects matching that name will be claimed.
```
Comment (nrb) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:216]: If ns is unset, wouldn't it be a cluster-scoped object? That's what I would expect from the rest of Kube.

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:216]: That's a very good question. Basically, yes, however in the authorizer we don't have any capability to check whether the request attributes reference a cluster-scoped object, hence my current implementation of the reverse permission authorizer doesn't make a difference and i don't see a similar one in RBAC authorizers from kube.

This is, however, somehow implied by the in-flight API Group/resource. But the invariant is not explicitely checked in the authorizer.

Conversely, one could set `name=foo` for a resource selector on a namespaced API Group/resource and the authorizer would reject any resource in any namespace matching that name.

This is how kube works as well. One can define a cluster role like:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: cm-reader
rules:
- apiGroups: [""]
  resources: ["configmaps"]
  resourceNames: ["foo"]
  verbs: ["get", "watch", "list"]
```

and bind that to users. Then, a user will only be able to read configmap `foo` in any namespace but will be denied to modify any of them.

**Comment by sttts** [comment]
/hold
/lgtm
/approve

@nrb are you fine to proceed? Seeing your comment. Still doubts?

**Comment by nrb** [comment]
Yeah, let's proceed.

/unhold

**Pull request merged**